### PR TITLE
Limit number of objects in mdag batch to prevent too many fds issue

### PR DESCRIPTION
**Pull request by whyrusleeping**
In certain situations the 8MB size limit here would result in a large number of objects in the batch, which combined with peer connections, can easily hit the file descriptor limit.

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

File: merkledag/merkledag.go [hunk 1]
```diff
@@ -68,7 +68,15 @@ func (n *dagService) Add(nd node.Node) (*cid.Cid, error) {
...
 func (n *dagService) Batch() *Batch {
-	return &Batch{ds: n, MaxSize: 8 * 1024 * 1024}
+	return &Batch{
+		ds:      n,
+		MaxSize: 8 << 20,
+
+		// By default, only batch up to 256 nodes at a time.
+		// The current implementation of flatfs opens this many file
+		// descriptors at the same time for the optimized batch write.
+		MaxBlocks: 256,
```
Comment (Kubuxu) [inline merkledag/merkledag.go]: I would reduce it even further. 256 blocks worst case scenario is 512 open fds (directory and file), and with almost 500 peers it is getting close to 1024 ulimit we set.

Comment (whyrusleeping) [inline merkledag/merkledag.go]: mmm... fair enough. How about 200 even? I think we should rework how flatfs works/replace it soon.

**Comment by whyrusleeping** [comment]
should resolve #3752

**Comment by Kubuxu** [comment]
It doesn't fix everything, I am still getting this failure:

```
expecting success: 
		ipfs add --raw-leaves -r -q somedir > /dev/null
	
22:51:36.165 ERROR commands/h: open /home/kubuxu/go/src/github.com/ipfs/go-ipfs/test/sharness/trash directory.t0270-filestore.sh/.ipfs/blocks/5H/put-659704546: too many open files client.go:247
Error: open /home/kubuxu/go/src/github.com/ipfs/go-ipfs/test/sharness/trash directory.t0270-filestore.sh/.ipfs/blocks/5H/put-659704546: too many open files
not ok 29 - normal add without fscache duplicates data
#	
#			ipfs add --raw-leaves -r -q somedir > /dev/null
#		
```

It is slightly modified test as previously the error was being hidden by var capture.

**Comment by Kubuxu** [comment]
Ok, now it vanished.

**Pull request merged**